### PR TITLE
[gapid2apk] add standalone APK creation to export_replay

### PR DESCRIPTION
**Pull request by hevrard**
This is all the code needed for gapid2apk on the gapir side: detect replay resources in the apk assets, and if present use them to do the replay.

The missing part is the external command to actually create a new APK with replay resources: I have this implemented as a bash script, but we will want it as a regular command in Go for the sake or portability.

I open the PR now for reviews of the GAPIR side of things, and I will add the APK creation command soon.

### CODE REVIEW

**Comment by ben-clayton** [comment]
Please ensure you squash these changes before submitting. Given that nobody has started reviewing yet, perhaps now would be a good time to squash to one or a few CLs?

File: cmd/gapit/trace2apk.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+	return nil
+}
+
+func fileCopy(dest, src string) error {
+	s, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+
```
Comment (ben-clayton) [inline cmd/gapit/trace2apk.go]: `defer s.Close()`

File: cmd/gapit/trace2apk.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	d, err := os.Create(dest)
+	if err != nil {
+		return err
+	}
+
```
Comment (ben-clayton) [inline cmd/gapit/trace2apk.go]: `defer d.Close()`

File: cmd/gapit/trace2apk.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	if err := client.ExportReplay(ctx, capturePath, nil, "assets/export_replay", &service.ExportReplayOptions{}); err != nil {
+		return log.Err(ctx, err, "Failed to export replay")
+	}
+
+	refAPKPath, err := layout.GapidApk(ctx, device.AndroidARM64v8a)
```
Comment (AWoloszyn) [inline cmd/gapit/trace2apk.go]: We probably don't want to hard-code the ABI here

File: cmd/gapit/trace2apk.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	// TODO(hevrard): implement the addition of assets, manifest package renaming (still todo!), re-sign, re-zip in core/os/android/apk module
+
+	// aapt add assets
+	execCommand(ctx, "aapt", "add", "-v", newAPK, "assets/export_replay/resources.data", "assets/export_replay/resources.index", "assets/export_replay/payload.bin")
+	user, err := user.Current()
+	if err != nil {
+		return err
+	}
+
+	keystore := filepath.Join(user.HomeDir, ".android/debug.keystore")
```
Comment (AWoloszyn) [inline cmd/gapit/trace2apk.go]: What should we do if this doesn't exist?
We may want to consider creating our own keystore if this is the case.

File: cmd/gapit/trace2apk.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	execCommand(ctx, "apksigner", "sign", "--ks", keystore, "--ks-pass", "pass:android", newAPK)
+
+	return nil
+}
+
+func fileCopy(dest, src string) error {
```
Comment (AWoloszyn) [inline cmd/gapit/trace2apk.go]: It looks like we have a similar function in core/os/file.go

File: core/cc/archive.cpp [hunk 6]
```diff
@@ -210,6 +210,10 @@ Archive::Archive(const std::string& archiveName)
     GAPID_FATAL("Unable to open archive index file %s", indexFilename.c_str());
   }
 
+  // Linux fopen() with mode "a" leads to reads from the beginning of file, but
+  // this is not true on Android, hence the explicit rewind() here
+  rewind(mIndexFile);
```
Comment (AWoloszyn) [inline core/cc/archive.cpp]: This seems counter to what people will expect, we may want to file a bug.

File: cmd/gapit/trace2apk.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+	"github.com/google/gapid/core/log"
+)
+
+type trace2apkVerb struct{ Trace2apkFlags }
+
+func init() {
+	verb := &trace2apkVerb{}
+	app.AddVerb(&app.Verb{
+		Name:      "trace2apk",
+		ShortHelp: "Create an stand-alone APK which replays a trace (off-screen)",
```
Comment (pau-baiget) [inline cmd/gapit/trace2apk.go]: Create an -> Create a

**Comment by AWoloszyn** [comment]
FYI: This may be a future update, but I tried this, and the performance was rather abysmal. From what I could tell, almost all of the time was spend unzipping resources from the archive. (Which is somewhat expected).

File: cmd/gapit/trace2apk.go [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+		return err
+	}
+
+	trace := filepath.Base(flags.Arg(0))
+	replayLabel := strings.TrimSuffix(trace, ".gfxtrace")
+	replayPackage := replayLabel + ".replay"
+	replayAPK := replayPackage + ".apk"
+	log.I(ctx, "Replay apk: %s", replayAPK)
+
+	// Extract reference APK
```
Comment (AWoloszyn) [inline cmd/gapit/trace2apk.go]: We should consider unzipping ourselves, rather than relying on the "unzip" command-line tool. Which won't exist on windows. Golang has this functionality more or less built in.
https://golangcode.com/unzip-files-in-go/

File: cmd/gapit/trace2apk.go [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+	type ManifestEntries struct {
+		Label   string
+		Package string
+	}
+	var entries = ManifestEntries{replayLabel, replayPackage}
+	t.Execute(f, entries)
+	f.Close()
+
+	// Re-assemble new APK with the new manifest
+	if err := execCommand(ctx, "aapt2", "link", "-o", replayAPK, "--manifest", "NewManifest.xml", "-I", os.ExpandEnv("${ANDROID_SDK_HOME}/platforms/android-26/android.jar")); err != nil {
```
Comment (AWoloszyn) [inline cmd/gapit/trace2apk.go]: It may make sense to use the aapt2 from the Android SDK. That way more users will have it work.

Comment (AWoloszyn) [reply cmd/gapit/trace2apk.go]: `${ANDROID_SDK_HOME}/platforms/android-26/android.jar` This probably won't work on Windows.

We can `path.join(getenv(ANDROID_SDK_HOME),  ...)`
Also it would be a good idea allow the user specify an sdk directory that isn't ANDROID_SDK_HOME. Forcing env_vars is not nice.

File: cmd/gapit/trace2apk.go [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+	}
+
+	// Add replay assets, uncompressed
+	assets, err := ioutil.ReadDir("assets/replay_export/")
+	if err != nil {
+		return err
+	}
+	for _, a := range assets {
+		// Arguments ("-0", "") ensure the asset will not be compressed
+		if err := execCommand(ctx, "aapt", "add", "-0", "", replayAPK, "assets/replay_export/"+a.Name()); err != nil {
```
Comment (AWoloszyn) [inline cmd/gapit/trace2apk.go]: Same as above with aapt.

File: cmd/gapit/trace2apk.go [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+		}
+	}
+
+	// Add the replay libraries
+	// TODO(#2677): get rid of this workaround once ABI naming is sanitized
+	abi := capture.ABI.Name
+	if abi == "armeabi" {
+		abi = "armeabi-v7a"
+	}
+	if err := execCommand(ctx, "zip", "-u", replayAPK, "lib/"+abi+"/libgapir.so", "lib/"+abi+"/libVkLayer_VirtualSwapchain.so"); err != nil {
```
Comment (AWoloszyn) [inline cmd/gapit/trace2apk.go]: We should be able to zip directly from golang.

File: cmd/gapit/export_replay.go [hunk 12]
```diff
@@ -98,5 +114,155 @@ func (verb *exportReplayVerb) Run(ctx context.Context, flags flag.FlagSet) error
...
+		}
+		capture := boxedCapture.(*service.Capture)
+
+		refAPKPath, err := layout.GapidApk(ctx, capture.ABI)
+		if err != nil {
+			return err
+		}
+
+		// Extract reference APK
+		if err := execCommand(ctx, "unzip", "-o", refAPKPath.String()); err != nil {
```
Comment (AWoloszyn) [inline cmd/gapit/export_replay.go]: I will re-add these comments in case they got buried from the previous MRs.

We should avoid using command-line executables that we cannot guarantee exist. zip/unzip are not present on windows for example.

File: cmd/gapit/export_replay.go [hunk 13]
```diff
@@ -98,5 +114,155 @@ func (verb *exportReplayVerb) Run(ctx context.Context, flags flag.FlagSet) error
...
+		}
+		type ManifestEntries struct {
+			Package string
+		}
+		var entries = ManifestEntries{replayPackage}
+		t.Execute(f, entries)
+		f.Close()
+
+		// Re-assemble new APK with the new manifest
+		if err := execCommand(ctx, "aapt2", "link", "-o", verb.Apk, "--manifest", "NewManifest.xml", "-I", os.ExpandEnv("${ANDROID_SDK_HOME}/platforms/android-26/android.jar")); err != nil {
```
Comment (AWoloszyn) [inline cmd/gapit/export_replay.go]: We should use aapt2 from the ANDROID_SDK_HOME.
Also it would be good to be able to specify the android_sdk_home from the command-line as well.

`--sdk_root="foo"`.  If not set then use the env var.

Comment (AWoloszyn) [inline cmd/gapit/export_replay.go]: Does this HAVE To be android-26/android.jar?
Can this be a different jar file? If it must be 26, that's fine I just want to make sure we are flexible if we can be.

File: cmd/gapit/export_replay.go [hunk 14]
```diff
@@ -98,5 +114,155 @@ func (verb *exportReplayVerb) Run(ctx context.Context, flags flag.FlagSet) error
...
+		}
+
+		// Add replay assets, uncompressed
+		assets, err := ioutil.ReadDir("assets/replay_export/")
+		if err != nil {
+			return err
+		}
+		for _, a := range assets {
+			// Arguments ("-0", "") ensure the asset will not be compressed
+			if err := execCommand(ctx, "aapt", "add", "-0", "", replayAPK, "assets/replay_export/"+a.Name()); err != nil {
```
Comment (AWoloszyn) [inline cmd/gapit/export_replay.go]: Same as above, we should use aapt2 from the SDK `build-tools/**/aapt2`

File: cmd/gapit/export_replay.go [hunk 15]
```diff
@@ -98,5 +114,155 @@ func (verb *exportReplayVerb) Run(ctx context.Context, flags flag.FlagSet) error
...
+			}
+		}
+
+		// Add the replay libraries
+		// TODO(#2677): get rid of this workaround once ABI naming is sanitized
+		abi := capture.ABI.Name
+		if abi == "armeabi" {
+			abi = "armeabi-v7a"
+		}
+		if err := execCommand(ctx, "zip", "-u", replayAPK, "lib/"+abi+"/libgapir.so", "lib/"+abi+"/libVkLayer_VirtualSwapchain.so"); err != nil {
```
Comment (AWoloszyn) [inline cmd/gapit/export_replay.go]: We can zip from golang itself.

File: core/cc/archive.cpp (hunk 6 above)
Comment (hevrard) [reply core/cc/archive.cpp]: I filed a bug internally.

File: cmd/gapit/export_replay.go [hunk 16]
```diff
@@ -98,5 +114,155 @@ func (verb *exportReplayVerb) Run(ctx context.Context, flags flag.FlagSet) error
...
+			return err
+		}
+
+		// Add the compiled Java code
+		if err := execCommand(ctx, "zip", "-uj", replayAPK, "classes.dex"); err != nil {
+			return err
+		}
+
+		// Sign the new APK
+		if err := execCommand(ctx, "apksigner", "sign", "--ks", os.ExpandEnv("${HOME}/.android/debug.keystore"), "--ks-pass", "pass:android", replayAPK); err != nil {
```
Comment (AWoloszyn) [inline cmd/gapit/export_replay.go]: I dont think this env will work on Windows.
We should probably set up a dummy keystore if none exist.

Comment (AWoloszyn) [reply cmd/gapit/export_replay.go]: I think we need to run zipalign on the zi pfile before apksigner as well?

File: cmd/gapit/export_replay.go (hunk 13 above)
Comment (hevrard) [reply cmd/gapit/export_replay.go]: It has to match the targetSdkVersion in `gapidapk/android/apk/AndroidManifest.xml.in`, which is currently 26. I have abstracted this in export_replay, but not across export_replay and the manifest template.

**Comment by hevrard** [comment]
Thanks for the previous review, I tried to fix all remarks. This PR enables to create a stand-alone APK which replay a given trace. We are still facing issues when trying to trace the created APKs, but we can fix these issues later.

As it is a big change, we may think of splitting this in a few PRs, feel free to suggest such splitting.

File: cmd/gapit/export_replay.go [hunk 17]
```diff
@@ -98,5 +127,293 @@ func (verb *exportReplayVerb) Run(ctx context.Context, flags flag.FlagSet) error
...
+		}
+		for _, a := range assets {
+			// Arguments ("-0", "") ensure the asset will not be compressed
+			if err := execCommand(ctx, aapt, "add", "-0", "", tmpAPK, path.Join(assetsPath, a.Name())); err != nil {
+				return err
+			}
+		}
+
+		// Add the replay libraries
+		// TODO(#2677): get rid of this workaround once ABI naming is sanitized
```
Comment (pmuetschard) [inline cmd/gapit/export_replay.go]: this is done, no?

File: cmd/gapit/export_replay.go [hunk 18]
```diff
@@ -98,5 +127,293 @@ func (verb *exportReplayVerb) Run(ctx context.Context, flags flag.FlagSet) error
...
+		if err := execCommand(ctx, apksigner, "sign", "--ks", keystorePath, "--ks-pass", "pass:android", replayAPKPath); err != nil {
+			return err
+		}
+
+	}
+
 	return nil
 }
+
+func execCommand(ctx context.Context, name string, args ...string) error {
```
Comment (pmuetschard) [inline cmd/gapit/export_replay.go]: We have core/os/shell that provides a nice API to do this sort of thing.

File: cmd/gapit/export_replay.go [hunk 19]
```diff
@@ -98,5 +127,293 @@ func (verb *exportReplayVerb) Run(ctx context.Context, flags flag.FlagSet) error
...
+	if err != nil {
+		return err
+	}
+	defer z.Close()
+
+	for _, f := range z.File {
+		destPath := filepath.Join(outputDir, f.Name)
+
+		// Protect from ZipSlip, see https://snyk.io/research/zip-slip-vulnerability
+		prefix := ""
```
Comment (pmuetschard) [inline cmd/gapit/export_replay.go]: The prefix can be determined outside of the loop.

File: cmd/gapit/export_replay.go [hunk 20]
```diff
@@ -98,5 +127,293 @@ func (verb *exportReplayVerb) Run(ctx context.Context, flags flag.FlagSet) error
...
+	out, err := cmd.CombinedOutput()
+	logger := log.From(ctx).Writer(log.Debug)
+	if err != nil {
+		logger = log.From(ctx).Writer(log.Error)
+	}
+	logger.Write(out)
+	return err
+}
+
+func unzip(ctx context.Context, archiveFilename string, outputDir string) error {
```
Comment (pmuetschard) [inline cmd/gapit/export_replay.go]: move this to core/os/file/zip.go

File: cmd/gapit/flags.go [hunk 21]
```diff
@@ -124,6 +124,8 @@ type (
 		OriginalDevice bool   `help:"export replay for the original device"`
 		Out            string `help:"output directory for commands and assets"`
 		OutputFrames   bool   `help:"generate trace that output frames(disable diagnostics)"`
+		Apk            string `help:"(experimental) create a stand-alone APK with that name which performs the replay"`
```
Comment (pmuetschard) [inline cmd/gapit/flags.go]: Please add here that the apk name should have the format "<app_package>.apk"

File: gapis/api/gles/api/egl.api [hunk 22]
```diff
@@ -357,14 +357,12 @@ cmd EGLBoolean eglSurfaceAttrib(EGLDisplay display,
   return ? // TODO
 }
 
-@no_replay
```
Comment (pmuetschard) [inline gapis/api/gles/api/egl.api]: Please move these API changes into their own commit with a detailed message of why we are making them.

**Comment by AWoloszyn** [comment]
Other than the comments from @pmuetschard it looks good to me.

File: cmd/gapir/cc/main.cpp [hunk 23]
```diff
@@ -341,35 +397,76 @@ void android_process(struct android_app* app, int32_t cmd) {
...
+
+  if (opts.mode == kReplayArchive) {
+    GAPID_INFO("Started Graphics API Replay from archive.");
+
+    waiting_thread = std::thread([&]() {
+      // It's important to use a different JNIEnv as it is a separate thread
+      JNIEnv* env;
+      app->activity->vm->AttachCurrentThread(&env, 0);
+
+      // Hugues: maybe we need to open assets in the main thread?
```
[the remark this answers is not in the document]
Comment (hevrard) [reply cmd/gapir/cc/main.cpp]: I clarified the comment, and thanks for the feedback on TODO formats!

File: cmd/gapit/export_replay.go (hunk 17 above)
Comment (hevrard) [reply cmd/gapit/export_replay.go]: yep!

File: cmd/gapit/export_replay.go (hunk 18 above)
Comment (hevrard) [reply cmd/gapit/export_replay.go]: thanks for the pointer

**Comment by hevrard** [comment]
Many thanks Pascal for the detailed feedback!
The EGL frame delimiters replay is now separated in #2708

**Comment by hevrard** [comment]
The OSX build fails with `ERROR: Aborting VM command due to timeout of 10800 seconds`, have you seen that before? How can we allocate more time for the build?
